httpcore/_backends/trio.py

```diff
@@ -55,12 +55,21 @@ async def start_tls(
 
     async def read(self, n: int, timeout: TimeoutDict) -> bytes:
         read_timeout = none_as_inf(timeout.get("read"))
-        exc_map = {trio.TooSlowError: ReadTimeout, trio.BrokenResourceError: ReadError}
+        exc_map = {trio.TooSlowError: ReadTimeout, OSError: ReadError}
 
         async with self.read_lock:
             with map_exceptions(exc_map):
                 with trio.fail_after(read_timeout):
-                    return await self.stream.receive_some(max_bytes=n)
+                    try:
+                        data = await self.stream.receive_some(max_bytes=n)
+                    except trio.BrokenResourceError:
+                        message = "Server disconnected while attempting read"
+                        raise OSError(message) from None
+
+                    if data == b"":
+                        raise OSError("Server disconnected while attempting read")
```

Comment:
Interestingly, `trio` raises an exception in the HTTP/2 case and returns `b""` in the HTTP/1.1 case (causing a `ProtocolError` from `h11`). I'm not sure why, because in both cases regular TCP sockets are used and we attempt reading 4096 bytes. So the solution I found is to deal with both possibilities…